Azure Architecture guidance (#6140)
Particular/docs.particular.net: architecture/azure/messaging.md

:heavy_plus_sign: Pros:

- Up to 100 MB message size on the Premium tier
- Supports transactions across multiple queues
- Scripted infrastructure deployment using the [NServiceBus ASB CLI](/transports/azure-service-bus/operational-scripting.md)
- Supports the AMQP 1.0 protocol
- Supports cross-entity transactions

:heavy_minus_sign: Cons:

- Some features are only available on the Premium tier
- The maximum message size is 256 KB on the Standard tier
- The maximum queue size is 80 GB on the Standard tier
- There is no on-premises equivalent.

[**Try the Azure Service Bus Tutorial →**](https://learn.microsoft.com/en-us/azure/service-bus-messaging/build-message-driven-apps-nservicebus?tabs=Sender)

#### When to use Azure Service Bus

Azure Service Bus may be considered the default messaging choice for Azure. Alternatives should be considered only when the message size limitations are not sufficient or if a more portable queueing technology is required.

> ### Azure Queue Storage

Reviewer: Funny that MS doesn't even mention on their page about messaging https://learn.microsoft.com/en-us/azure/architecture/guide/technology-choices/messaging

Author: it's extremely hard to find documentation for Azure Storage at this point. Maybe it's worthwile raising this as an observation

Author: FYI commented on https://github.com/Particular/GeneralPlatformExperience/issues/1485

Reviewer 2: > it's extremely hard to find documentation for Azure Storage at this point. Maybe it's worthwile raising this as an observation

That's because it's a legacy messaging service that is not promoted.

### Azure Service Bus

[Azure Service Bus](https://learn.microsoft.com/en-us/azure/service-bus-messaging/service-bus-messaging-overview) supports queuing, publish/subscribe, and more advanced integration patterns. It is designed to integrate applications or application components that may span multiple communication protocols, data contracts, trust domains, or network environments.

Azure Service Bus has two pricing tiers: [Standard and Premium](https://learn.microsoft.com/en-us/azure/service-bus-messaging/service-bus-premium-messaging). The Premium tier provides hardware level isolation, which means each customer's workload runs on dedicated hardware. Premium tier uses _messaging units_ for pricing. The Standard tier offers pay-as-you-go pricing with less predictable latency and throughput. Microsoft recommends running production systems on the Premium tier.

:heavy_plus_sign: Pros:

- Up to 100 MB message size on the Premium tier

> - Supports transactions across multiple queues

Reviewer 2: Nice please - put a footnote to emphasize that the service supports messaging transactions, not multi-service transactions.

Reviewer 2: > to avoid a footnote, is `Supports transactional message operations across multiple queues` a better wording?

Almost there. Technically speaking, it's not just queues. You could cover queues and topics (for pub/sub) with
> Supports transactional message operations across multiple entities

or 

> Supports transactional message operations across multiple queues and topics

> - Supports transactions across multiple queues

Reviewer 2: ```suggestion
- Supports cross-entity transactions
```

- Scripted infrastructure deployment using the [NServiceBus ASB CLI](/transports/azure-service-bus/operational-scripting.md)

> - Supports the AMQP 1.0 protocol

Reviewer 2: ```suggestion
- Supports the AMQP 1.0 protocol over TCP and WebSockets
```

> - Supports cross-entity transactions

Reviewer 2: Already mentioned above
```suggestion
```

:heavy_minus_sign: Cons:

> - Some features are only available on the Premium tier

Reviewer 2: ```suggestion
- Some features are only available on the Premium tier
- Cross-entity transactions are limited to 100 messages
- No cross-namespace communication
```

- The maximum message size is 256 KB on the Standard tier
- The maximum queue size is 80 GB on the Standard tier

> - There is no on-premises equivalent.

Reviewer 2: This is under the Azure section for systems that will be built in the cloud. Don't think anyone would expect to be able to run their production using ASB on-premises, right?
If you mention there's no emulator for local development, that's fine.

```suggestion
```

[**Try the Azure Service Bus Tutorial →**](https://learn.microsoft.com/en-us/azure/service-bus-messaging/build-message-driven-apps-nservicebus?tabs=Sender)

#### When to use Azure Service Bus

> 

Reviewer 2: I'd suggest steering the reader from using ASQ.

Reviewer 2: Web/worker communication was originally the primary use case for ASQ some 10+ years ago.

There's no official mention of any sort. And there's also no development of the service. And it's also not really listed as a [messaging service](https://learn.microsoft.com/en-us/azure/service-bus-messaging/compare-messaging-services) either. It is similar to Storage Tables. With one exception, perhaps - Storage Tables API could be switched to Cosmos DB Table API, and the systems would still work. With Storage Queues, there's no such option (read no team that took it upon themselves).

A service MSFT wishes wouldn't be around anymore, but it must keep the lights on because it's in many legacy systems.

Azure Service Bus may be considered the default messaging choice for Azure. Alternatives should be considered only when the message size limitations are not sufficient or if a more portable queueing technology is required.

### Azure Queue Storage

[Azure Queue Storage](https://learn.microsoft.com/en-us/azure/storage/queues/storage-queues-introduction) offer a simple, low-cost, and easy-to-use messaging service based on the Azure Storage infrastructure.

:heavy_plus_sign: Pros:

- Queues can contain up to 200 TB of messages
- Cheaper than Azure Service Bus
- Part of Azure Storage

> - The Azure Storage Emulator may be on-premises for development, testing, and CI

Reviewer 2: ```suggestion
- The Azure Storage Emulator for development and testing
```

:heavy_minus_sign: Cons:

> - The maximum message size is 64 KB
> - Higher latency due to HTTP communication
> - No cross-entity transactions

Reviewer 2: ```suggestion
- The maximum message size is 64 KB
- Higher latency due to HTTP communication
- No native support for headers
- No cross-entity transactions
```

Author: given we fill in the support for headers, I'm not sure that is something super important to mention here.

- The maximum number of requests (e.g. message delivery) per storage account is 20,000 per second (2,000 per second for single queues)

[**Try the Azure Storage Queues sample →**](/samples/azure/storage-queues/)

#### When to use Azure Storage Queues

> Consider Azure Storage Queues when a managed message queuing technology is required but Azure Service Bus doesn't meet requirements or is too expensive.

Reviewer 2: ```suggestion
Azure Storage Queues is a legacy Azure service that is not recommended for new systems.
```

$10/m for the ASB standard tier for a commercial system running on NSB is cheaper than the licensing cost of NSB for a solution that includes 4+ endpoints, which is any NSB-based system in production.

Reviewer 2: I've updated the suggestion. Have a look

### SQL transport

SQL transport is an NServiceBus feature that can use existing SQL Server-compatible databases as feature-complete message queues.

:heavy_plus_sign: Pros:

- Runs on cloud hosted and on-premises SQL Server-compatible data stores (including SQL Server Express for development, testing, and CI)
- Arbitrary message sizes
- Runs on infrastructure which often already exists

> - Strong transaction integration with business data operations

Reviewer 2: I'd call this out. This is one of the selling points.

The other one is the fact that no new service needs to be provisioned to enable messaging.

Reviewer 2: There's another one, but I'm not sure you'd want to list it - saga timeouts are queriable 🙈

:heavy_minus_sign: Cons:

- More expensive and laborious to scale
- Impacts overall database performance
- Lower message throughput compared to specialized message queuing technologies

[**Try the SQL transport sample →**](/samples/azure/storage-queues/)

#### When to use SQL transport

> Consider SQL transport if an existing application already uses a SQL Server-compatible data store and limited amount of messaging is being introduced. SQL transport can be a good stepping-stone when introducing messaging into an existing system without the introduction of new infrastructure.

Reviewer 2: 👍

## Other services

Azure offers other messaging services focused on asynchronous communication. These services are not directly support by the Particular Platform, however they might be combined in combination with supported message queueing systems to handle [data distribution](/nservicebus/concepts/data-distribution.md) needs by more specialized technologies.

> * [Azure Event Grid](https://learn.microsoft.com/en-us/azure/event-grid/overview) is an integration focused messaging system using the publish-subscribe model. Azure Event Grid can ingest and distribute events provided by applications, other Azure Services or IoT devices using HTTP protocols (MQTT support in public preview). By default, Event Grid uses a [push-model](https://learn.microsoft.com/en-us/azure/event-grid/push-delivery-overview) to forward events to all subscribers.

Reviewer 2: With Event Grid v2 this needs to change IMO.

Ideally, NSB should have a transport EG with [pull-based delivery](https://learn.microsoft.com/en-us/azure/event-grid/pull-delivery-overview). It's not as polished and rounded as ASB but it will be there. And for customers that are more cloud-native/first, this transport would be a natural option.

Author: I'll make an internal note to keep this in mind and update once the new APIs move out of preview

---
title: Azure messaging
summary:
reviewed: 2023-07-18
callsToAction: ['solution-architect', 'poc-help']
---

> Azure offers [several messaging services](https://learn.microsoft.com/en-us/azure/service-bus-messaging/compare-messaging-services), each built for specific purposes. The Particular Service Platform supports [Azure Service Bus and Azure Queue Storage](https://learn.microsoft.com/en-us/azure/service-bus-messaging/service-bus-azure-and-service-bus-queues-compared-contrasted).

Reviewer 3: Article linked calls them "storage queues" - any reason why this is in the opposite order?

Author: yeah microsoft doesn't seem to be consistent with it either, e.g. https://learn.microsoft.com/en-us/azure/storage/queues/storage-queues-introduction calls it Azure Queue Storage 🤷

### Azure Service Bus

[Azure Service Bus](https://learn.microsoft.com/en-us/azure/service-bus-messaging/service-bus-messaging-overview) is a fully-fledged enterprise messaging service with support for queuing, publish/subscribe, and more advanced integration patterns. It is designed to integrate applications or application components that may span multiple communication protocols, data contracts, trust domains, or network environments.

Azure Service Bus has two pricing tiers: [Standard and Premium](https://learn.microsoft.com/en-us/azure/service-bus-messaging/service-bus-premium-messaging). The Premium tier provides hardware-level isolation, which means each customer's workload runs on dedicated hardware. Premium tier uses _messaging units_ for scaling and pricing and guarantees predictable throughput and performance. The Standard tier offers pay-as-you-go pricing with quota-based throughput and non-guaranteed latency. Microsoft recommends running production systems on the Premium tier.

:heavy_plus_sign: Pros:

> - Up to 100 MB message size on the Premium tier

Reviewer 3: This feels like we're just regurgitating Azure marketing stuff. Should we mention how large messages are generally an anti-pattern, mention the claim check pattern, and/or link somehow to data bus stuff?

Author: I agree that large messages should be considered an anti-pattern, but I'm not sure that the claim-check pattern would be the default response to that (rather than revisiting the message contracts and service boundaries etc.).

At least on the command side, I'm not sure why we would strongly recommend going with a claim-check pattern and an external data store when the messaging system already covers that functionality?

> - Supports transactional message operations across multiple queues and topics (cross-entity transactions)

Reviewer 3: This feels like it _should_ be the biggest Pro.

Reviewer 2: And the biggest con

Author: yeah, will move this up.

- Scripted infrastructure deployment using the [NServiceBus ASB CLI](/transports/azure-service-bus/operational-scripting.md)
- Supports the AMQP 1.0 protocol over TCP and WebSockets

:heavy_minus_sign: Cons:

- Some features are only available on the Premium tier
- The maximum message size is 256 KB on the Standard tier

> - The maximum queue size is 80 GB on the Standard tier

Reviewer 3: Azure and us say you shouldn't use Standard for production, so why should the limitations of Standard be a Con consideration?

- Cross-entity transactions are limited to 100 messages
- No cross-namespace communication
- No emulator for local development or testing available

[**Try the Azure Service Bus Tutorial →**](https://learn.microsoft.com/en-us/azure/service-bus-messaging/build-message-driven-apps-nservicebus?tabs=Sender)

#### When to use Azure Service Bus

> Azure Service Bus may be considered the default messaging choice for Azure. Alternatives should be considered only when the message size limitations are not sufficient or if a more portable queueing technology is required.

Reviewer 3: ```suggestion
Azure Service Bus should be considered the default messaging choice for Azure. Alternatives should be considered only when the message size limitations are not sufficient or if a more portable queueing technology is required.
```

> Azure Service Bus may be considered the default messaging choice for Azure. Alternatives should be considered only when the message size limitations are not sufficient or if a more portable queueing technology is required.

Reviewer 3: I'd be tempted to go further than my minor edit above and remove the bit about message size restrictions and maybe the thing about portability too. IMO, the only reason to go ASQ is if you're cash-strapped.

### Azure Queue Storage

> [Azure Queue Storage](https://learn.microsoft.com/en-us/azure/storage/queues/storage-queues-introduction) offer a simple, low-cost, and easy-to-use messaging service based on the Azure Storage infrastructure.

Reviewer 3: ```suggestion
[Azure Queue Storage](https://learn.microsoft.com/en-us/azure/storage/queues/storage-queues-introduction) offers a simple, low-cost, and easy-to-use messaging service based on the Azure Storage infrastructure.
```

Reviewer 3: I'd be tempted to add "but without a lot of the nicer features of Azure Service Bus" or something like that. Again, are we only mimicking Azure marketing points here?

Author: maybe "[...] easy-to-use **minimal** messaging service"?

:heavy_plus_sign: Pros:

- Queues can contain up to 200 TB of messages
- Cheaper than Azure Service Bus

> - Part of Azure Storage

Reviewer 3: I'd explain why that qualifies as a Pro or leave it out, as it's not self-evident

- Azure Storage Emulator for local development and testing

:heavy_minus_sign: Cons:

- The maximum message size is 64 KB
- Higher latency due to HTTP communication
- No cross-entity transactions
- The maximum number of requests (e.g. message delivery) per storage account is 20,000 per second (2,000 per second for single queues)

[**Try the Azure Storage Queues sample →**](/samples/azure/storage-queues/)

#### When to use Azure Storage Queues

> Consider Azure Storage Queues when a managed message queuing technology is required for simple messaging needs or when Azure Service Bus doesn't meet critical requirements.

Reviewer 3: > Consider Azure Storage Queues when a managed message queuing technology is required for simple messaging needs

That's valid for ASB too?

> or when Azure Service Bus doesn't meet critical requirements

That's…super vague.

Author: > That's…super vague.

intentionally, because we already state that we recommend ASB by default (in the ASB section). I guess this could be repeated.

### SQL transport

> SQL transport is an NServiceBus feature that can use existing SQL Server-compatible databases as feature-complete message queues.

Reviewer 3: What does SQL Server **compatible** mean?

Author: hm not sure what the idea was, probably safe to remove.

:heavy_plus_sign: Pros:

- Runs on infrastructure which often already exists
- Strong transaction integration with business data operations
- Runs on cloud hosted and on-premises SQL Server-compatible data stores (including SQL Server Express for development, testing, and CI)
- Arbitrary message sizes

:heavy_minus_sign: Cons:

- More expensive and laborious to scale
- Impacts overall database performance
- Lower message throughput compared to specialized message queuing technologies

[**Try the SQL transport sample →**](/samples/azure/storage-queues/)

#### When to use SQL transport

Consider SQL transport if an existing application already uses a SQL Server-compatible data store and limited amount of messaging is being introduced. SQL transport can be a good stepping-stone when introducing messaging into an existing system without the introduction of new infrastructure.

## Other services

> Azure offers other messaging services focused on asynchronous communication. These services are not directly support by the Particular Platform, however they might be combined in combination with supported message queueing systems to handle [data distribution](/nservicebus/concepts/data-distribution.md) needs by more specialized technologies.

Reviewer 3: ```suggestion
Azure offers other messaging services focused on asynchronous communication. These services are not directly supported by the Particular Service Platform, however they might be combined in combination with supported message queueing systems to handle [data distribution](/nservicebus/concepts/data-distribution.md) needs by more specialized technologies.
```
